### ISSUE TITLE
Breaking change: Exception on adding custom claim type that already exists on token

### ISSUE THREAD

**Issue by ace90210** [DuendeSoftware/IdentityServer#545]
**Which version of Duende IdentityServer are you using?**
 Duende IdentityServer 5.2.2
 
 Previously worked in Identity Server 3.1.3 and 4.1.2

**Which version of .NET are you using?**
Core 3.1

**Describe the bug**
When adding a custom claim that conflicts with existing claims on the token an exception occurs on line 109 of the TokenExtension.cs. This appears to be due to the fact previously the general "AddClaims" method was used which explicitly handles duplicate claims (by making them an array when detected). This was changed to manually checking and adding claims. The new approach does not check if the claim type is already in the dictionary before trying to add it.

**To Reproduce**

1. Create a client with client credentials flow and the custom claim below

2. `Claims = new List<ClientClaim>() { new ClientClaim("aud", "mycustomaudiencecheck") }`
 
3. request a token for this client and receive a 500 response

**Expected behaviour**
The same as previous behaviour. The existing claim is either converted to an array (if it was not an array already) and then the claim is added or the claim is added to the existing array of that type.

**Log output/exception with stacktrace**

```
System.ArgumentException: An item with the same key has already been added. Key: aud
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at Duende.IdentityServer.Extensions.TokenExtensions.CreateJwtPayloadDictionary(Token token, IdentityServerOptions options, ISystemClock clock, ILogger logger) in /_/src/IdentityServer/Extensions/TokenExtensions.cs:line 109
   at Duende.IdentityServer.Services.DefaultTokenCreationService.CreatePayloadAsync(Token token) in /_/src/IdentityServer/Services/Default/DefaultTokenCreationService.cs:line 85
   at Duende.IdentityServer.Services.DefaultTokenCreationService.CreateTokenAsync(Token token) in /_/src/IdentityServer/Services/Default/DefaultTokenCreationService.cs:line 76
   at Duende.IdentityServer.Services.DefaultTokenService.CreateSecurityTokenAsync(Token token) in /_/src/IdentityServer/Services/Default/DefaultTokenService.cs:line 292
   at Duende.IdentityServer.ResponseHandling.TokenResponseGenerator.CreateAccessTokenAsync(ValidatedTokenRequest request) in /_/src/IdentityServer/ResponseHandling/Default/TokenResponseGenerator.cs:line 426
   at Duende.IdentityServer.ResponseHandling.TokenResponseGenerator.ProcessTokenRequestAsync(TokenRequestValidationResult validationResult) in /_/src/IdentityServer/ResponseHandling/Default/TokenResponseGenerator.cs:line 342
   at Duende.IdentityServer.ResponseHandling.TokenResponseGenerator.ProcessAsync(TokenRequestValidationResult request) in /_/src/IdentityServer/ResponseHandling/Default/TokenResponseGenerator.cs:line 104
   at Duende.IdentityServer.Endpoints.TokenEndpoint.ProcessTokenRequestAsync(HttpContext context) in /_/src/IdentityServer/Endpoints/TokenEndpoint.cs:line 106
   at Duende.IdentityServer.Endpoints.TokenEndpoint.ProcessAsync(HttpContext context) in /_/src/IdentityServer/Endpoints/TokenEndpoint.cs:line 71
   at Duende.IdentityServer.Hosting.IdentityServerMiddleware.Invoke(HttpContext context, IEndpointRouter router, IUserSession session, IEventService events, IIssuerNameService issuerNameService, IBackChannelLogoutService backChannelLogoutService) in /_/src/IdentityServer/Hosting/IdentityServerMiddleware.cs:line 103
   at Duende.IdentityServer.Hosting.IdentityServerMiddleware.Invoke(HttpContext context, IEndpointRouter router, IUserSession session, IEventService events, IIssuerNameService issuerNameService, IBackChannelLogoutService backChannelLogoutService) in /_/src/IdentityServer/Hosting/IdentityServerMiddleware.cs:line 103
   at Duende.IdentityServer.Hosting.MutualTlsEndpointMiddleware.Invoke(HttpContext context, IAuthenticationSchemeProvider schemes) in /_/src/IdentityServer/Hosting/MutualTlsEndpointMiddleware.cs:line 95
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Duende.IdentityServer.Hosting.DynamicProviders.DynamicSchemeAuthenticationMiddleware.Invoke(HttpContext context) in /_/src/IdentityServer/Hosting/DynamicProviders/DynamicSchemes/DynamicSchemeAuthenticationMiddleware.cs:line 48
   at Duende.IdentityServer.Hosting.BaseUrlMiddleware.Invoke(HttpContext context) in /_/src/IdentityServer/Hosting/BaseUrlMiddleware.cs:line 32
   at Microsoft.AspNetCore.Localization.RequestLocalizationMiddleware.Invoke(HttpContext context)
```

**Additional context**

This was spotted when an existing large project we run updated IDS to Duende 5.2.2 and this bug was not spotted until it was too late. One of our consumers did not want to add the full scope access to a resource but wanted to check it on the audience instead. 

They achieved this in our previous IDS by adding a custom client claim (type "aud", Value "customservicename"). After the update when trying to get a Client Credentials token they got a 500 response and reported it to us. We checked to logs and traced the issue to the above.

**Comment by InKolev** [comment]
We experience the same behaviour when we updated from **IDS V3** to **Duende V5**.

It throws the "**System.ArgumentException: An item with the same key has already been added.** " when we use a reserved claim word like **"aud"**. Previously, the custom "aud" value was pushed to the existing "aud" claims array and no errors were thrown.

Was this intentionally changed or is it an accidental breaking change?

**Comment by leastprivilege** [comment]
We will look into it, thanks.

Generally speaking we treat the audience claims differently, since the content of the aud claim should be determined by the granted API resources that are requested. Manually adding an aud claims seems a bit "hacky". Nevertheless - we will investigate.

**Comment by brockallen** [comment]
Can you explain in more detail why you need to override the aud claim? Why does our ApiResource model not work for you? THanks

**Comment by ace90210** [comment]
The consumers using the aud claim override didnt say the justification, once we raised this bug with the and suggested the alternative options they have chosen to just use the scope to resolve their issue.

We don't ourselves need a fix for this anymore as they are happy to resolve it this way but as its technically a breaking change and others may find a legitimate need for adding it , it might still be worth resolving.

As part of my  research into the bug i did clone down Identity Server and fix the bug myself just to see if it could be one without too much hassle and was able too. If you want i could PR the fix i implemented however its only been tested on this one case.

**Comment by brockallen** [comment]
We discussed this, and decided that the new design is better, as there are certain token claims where it's important that we control them and claims coming from elsewhere (like the profile service) should not interfere. But, if you really do need to override this behavior, you can, it's just with a different (and more targeted) extensibility point. Thanks.

**Comment by brockallen** [comment]
Just an update on this -- we did in fact make a change to address this issue. We do add special claims that have to do with the token validation, and so extra claims that collide with those claims are now ignored (as opposed to us letting the Dictionary throw):

https://github.com/DuendeSoftware/IdentityServer/pull/633

And we released a 5.2.4 patch for this today. Thanks.
